selfhost/typecheck.wyv

```diff
@@ -7,14 +7,34 @@ import wyvern.collections.llist
 type List = llist.LinkedList
 type DeclType = types.DeclType
 
-def unfoldType(x:types.NominalType, gamma:List[DeclType]):DeclType
-    val pred = ((b:DeclType) =>
-        val z = match b:
-            t:types.TypeType => types.equalBinding(x.L, t.name)
-            default => false
-        z
-    )
-    llist.find[DeclType](gamma, pred).get()
+/*
+def searchGamma[T](gamma:List[DeclType], pred:DeclType->Boolean):T
+    val opt = llist.find[DeclType](gamma, pred)
+    match opt:
+        s:option.Some => 
+            val ret:DeclType = opt.get()
+            match ret:
+                t:T => t
+                default => error.report("search in gamma failed",error.unknownLocation)
+        n:option.None => error.report("search in gamma failed",error.unknownLocation)
+*/
```

Comment:
this function doesn't work right now--I want to be able to return a subtype of DeclType since I will pass in a predicate function that will only select DeclTypes of that subtype. The issue is that in the innermost match, it doesn't know if T is a subtype of DeclType and so it won't work. Can this be modified so that this typechecks?